target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,461 @@
+ * clocks {
+ *   osc: oscillator {
+ *     compatible = "fixed-clock";
+ *     #clock-cells = <0>;
+ *     clock-frequency = <25000000>;
+ *   };
+ *   ccu: clock-controller {
+ *     compatible = "realtek,rtl838x-clock";
+ *     #clock-cells = <1>;
+ *     clocks = <&osc>;
+ *     clock-names = "ref_clk";
+ *   };
+ * };
+ *
+ * Take care to set the compatible value to match your device. E.g.
+ * either realtek,rtl838x-clock or realtek,rtl839x-clock. Afterwards
+ * one can access/identify the clocks in the DTS with <&ccu CLK_CPU>,
+ * <&ccu CLK_MEM> or <&ccu CLK_LXB>.
+ */
+
+#include <linux/clk-provider.h>
+#include <linux/clkdev.h>
+#include <linux/clk.h>
+#include <linux/cpu.h>
+#include <linux/io.h>
+#include <linux/ioport.h>
+#include <linux/of_address.h>
+#include <linux/slab.h>
+
+#include <dt-bindings/clock/rtl83xx-clk.h>
+#include <asm/mach-rtl838x/mach-rtl83xx.h>
+
+/*
+ * Hardware specific definitions
+ */
+
+static const int rtl838x_pll_base[] = {
+	RTL838X_PLL_CPU_CTRL0,
+	RTL838X_PLL_MEM_CTRL0,
+	RTL838X_PLL_LXB_CTRL0
+};
+
+static const int rtl839x_pll_base[] = {
+	RTL839X_PLL_CPU_CTRL0,
+	RTL839X_PLL_MEM_CTRL0,
+	RTL839X_PLL_LXB_CTRL0
+};
+
+static const int rtl83xx_pll_divn3[] = { 2, 3, 4, 6 };
+static const int rtl83xx_pll_xdiv[] = { 2, 4, 2 };
+
+/*
+ * Clock structures
+ */
+
+int rtl_soc_family;
+
+#define PLL_INFO(_id, _name, _pname)	\
+	{				\
+		.id = _id,		\
+		.name = _name,		\
+		.parent_name = _pname	\
+	}
+
+struct rtl_pll_info {
+	unsigned int id;
+	const char *name;
+	const char *parent_name;
+};
+
+struct rtl_pll {
+	struct clk_hw hw;
+	unsigned int id;
+	spinlock_t lock;
+};
+
+static const struct rtl_pll_info pll_info[] = {
+	PLL_INFO(CLK_CPU, "cpu_clk", "ref_clk"),
+	PLL_INFO(CLK_MEM, "mem_clk", "ref_clk"),
+	PLL_INFO(CLK_LXB, "lxb_clk", "ref_clk")
+};
+
+#define PLL_COUNT ARRAY_SIZE(pll_info)
+
+struct rtl_pll_data {
+	struct device_node *np;
+	struct rtl_pll *plls[PLL_COUNT];
+};
+
+struct rtl_pll_init_data {
+	unsigned int id;
+	const char *name;
+	const char *parent_name;
+	struct device_node *np;
+	const struct clk_ops *ops;
+};
+
+#define rtl_hw_to_pll(_hw) container_of(_hw, struct rtl_pll, hw)
+
+/*
+ * clock setter/getter functions of the driver
+ */
+
+static unsigned long rtl83xx_pll_recalc_rate(struct clk_hw *hw, unsigned long parent_rate)
+{
+	struct rtl_pll *pll = rtl_hw_to_pll(hw);
+	unsigned int pll_base = (rtl_soc_family == RTL8380_FAMILY_ID) ? rtl838x_pll_base[pll->id] : rtl839x_pll_base[pll->id];
+
+	unsigned int ctrl0 = sw_r32(pll_base);
+	unsigned int ctrl1 = sw_r32(pll_base + 4);
+	
+	unsigned int cmu_sel_prediv = ctrl0 & 0x3;
+	unsigned int cmu_sel_div4 = ctrl0 >> 2 & 0x1;
+	unsigned int cmu_ncode_in = ctrl0 >> 4 & 0xff;
+	unsigned int cmu_divn2 = ctrl0 >> 12 & 0xff;
+	unsigned int cmu_divn2_selb;
+	unsigned int cmu_divn3_sel;
+
+	if (rtl_soc_family == RTL8380_FAMILY_ID) {
+		cmu_divn2_selb = ctrl1 >> 26 & 1;
+		cmu_divn3_sel = ctrl1 >> 27 & 3;
+	} else {
```

Comment:
what else? what family do we have if its' NOT the RTL8380? better make it explicit with a switch case where the default/unknown case is an error.